### PR TITLE
Fix STRUCT comparison between Pandas and Spark dataframes in fastparquet tests

### LINKED ISSUE #9399: [BUG] Fix `STRUCT` comparison between Pandas and Spark dataframes in `fastparquet` tests
**Description**

This is a follow-up to NVIDIA/spark-rapids/pull/9366.

`STRUCT` row comparison between Pandas and Spark dataframes seems to fail, seemingly because of a notational difference.

Consider `fastparquet_compatibility_test.py::test_reading_file_written_by_spark_cpu`. For the `STRUCT` case, the differences look as follows:
```
--- CPU OUTPUT
+++ GPU OUTPUT
@@ -1 +1 @@
-Row(a.first=-341142443, a.second=3.333994866005594e-37)
+Row(a=Row(first=-341142443, second=3.333994866005594e-37))
```
The `Row` read from the GPU/Plugin contains a `Row`, which matches the behaviour of Apache Spark reads.  The same data when read through `fastparquet` and converted into a Spark dataframe does not have nested rows.

This test is `xfail`ed for `STRUCT`. It would be good to change the comparison code to account for this notational difference.

### PR DESCRIPTION
**Pull request by res-life**
closes #9399

- Add converter from fastparquet result to cpu type result
  The fastparquet reading parquet result is flatted, and is not the same as CPU type result.
  e.g.:  fastparquet Row(a.b1 = 1, a.b2 = 2) vs  CPU type Row(a = Row(b1 = 1, b2 = 2))
- Modify `assert_gpu_and_cpu_are_equal_collect` to add a converter function parameter.
- update the test case in https://github.com/NVIDIA/spark-rapids/pull/9366

### CODE REVIEW

**Review by revans2 — COMMENTED** [review]
Looks good to me.

File: integration_tests/src/main/python/asserts.py [hunk 1]
```diff
@@ -459,37 +459,39 @@ def do_it_all(spark):
...
+        return from_gpu
 
-    if is_cpu_first:
-        run_on_cpu()
-        run_on_gpu()
-    else:
-        run_on_gpu()
-        run_on_cpu()
-
+    (from_cpu, from_gpu) = (run_on_cpu(), run_on_gpu()) if is_cpu_first else (run_on_gpu(), run_on_cpu())
```
Comment (mythrocks) [inline integration_tests/src/main/python/asserts.py:485]: This is a little confusing to me.  The meaning of `is_cpu_first` used to be the choice of whether to execute the CPU run first.

On this line, it looks like if `is_cpu_first == False`, the GPU run happens first, but the result is assigned to `from_cpu`.  Am I reading this incorrectly?

File: integration_tests/src/main/python/fastparquet_utils.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    # recursively merge items to sub Row, from bottom to top.
+    # e.g.:
+    # step0: dict(a.b.c1 = 1, a.b.c2 = 2, a.b2 = 3)
+    # step1: dict(a.b = Row(c1 = 1, c2 = 2), a.b2 = 3)
+    # step2: dict(a = Row(b = Row(c1 = 1, c2 = 2), b2 = 3))
+    for gpu_k, gpu_v in gpu_dict.items():
+        if isinstance(gpu_v, Row):
+            cpu_prefix = cpu_prefix + gpu_k + "."
+            _merge_cpu_dict(cpu_dict, gpu_v.asDict(), cpu_prefix)
+            # redcue the dict for the `cpu_prefix`
```
Comment (mythrocks) [inline integration_tests/src/main/python/fastparquet_utils.py:61]: Nit:
```suggestion
            # reduce the dict for the `cpu_prefix`
```

**Comment by mythrocks** [comment]
Thanks for working on this, @res-life. I only had a couple of minor questions.

File: integration_tests/src/main/python/fastparquet_utils.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+def convert_fastparquet_result(cpu, gpu):
+    return [_convert_fastparquet_row(cpu_row, gpu_row) for (cpu_row, gpu_row) in zip(cpu, gpu)]
+
+def _convert_fastparquet_row(cpu_row, gpu_row):
+    cpu_dict = cpu_row.asDict()
+    _merge_cpu_dict(cpu_dict, gpu_row.asDict())
+    return Row(**cpu_dict)
+
+def _merge_cpu_dict(cpu_dict, gpu_dict, cpu_prefix=""):
```
Comment (winningsix) [inline integration_tests/src/main/python/fastparquet_utils.py:51]: Hmm, is ```gpu_dict``` a must parameter to pass in? It seems we can directly build from cpu_dict. It's more like a flatted representation of GPU format.

File: integration_tests/src/main/python/fastparquet_utils.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    prefix_key = cpu_prefix[0 : len(cpu_prefix)-1]
+    # set the merge row
+    cpu_dict[prefix_key] = row
+
+
+def _test_convert_fastparquet_result():
+    """
+    verify the method `_convert_fastparquet_result`
+    """
+    cpu_dict = {'v1_1' : 11, 'v1_2.v2_1' : 21, 'v1_2.v2_2.v3_1' : 31, 'v1_2.v2_2.v3_2' : 32}
```
Comment (winningsix) [inline integration_tests/src/main/python/fastparquet_utils.py:93]: Just to understand null case: If we have a null value, will schema for both side stay the same? e.g., like v1_1 has no value. what's the value to represent null in both sides?

We probably need to add 1 more test case contains null in case different schema or null value representation.

File: integration_tests/src/main/python/asserts.py (hunk 1 above)
Comment (winningsix) [reply integration_tests/src/main/python/asserts.py:485]: Hmm, same question since the order matters in ```result_converter_func_before_compare```.

File: integration_tests/src/main/python/asserts.py [hunk 5]
```diff
@@ -459,37 +459,39 @@ def do_it_all(spark):
...
     print('### {}: GPU TOOK {} CPU TOOK {} ###'.format(collect_type,
         gpu_end - gpu_start, cpu_end - cpu_start))
+
+    if result_converter_func_before_compare is not None:
+        (from_cpu, from_gpu) = result_converter_func_before_compare(from_cpu, from_gpu)
+        # after the above conversion, the result is changed,
+        # so should sort result locally before compare, add @ignore_order(local=True) to the test case
+        assert should_sort_locally()
```
[the remark this answers is not in the document]
Comment (firestarman) [reply integration_tests/src/main/python/asserts.py:493]: > `# after the above conversion, the result is changed,` 

This is true for the fastparquet case here, but may be false for some cases. We'd better not enforce the local sort. Callers should add it when they know it is necessary.

File: integration_tests/src/main/python/asserts.py [hunk 6]
```diff
@@ -557,13 +559,23 @@ def run_on_gpu():
 
     return (from_cpu, from_gpu)
 
-def assert_gpu_and_cpu_are_equal_collect(func, conf={}, is_cpu_first=True):
+def assert_gpu_and_cpu_are_equal_collect(func, conf={}, is_cpu_first=True, result_converter_func_before_compare=None):
```
Comment (firestarman) [inline integration_tests/src/main/python/asserts.py:562]: nit: How about give a shorter name, e.g. `post_handle_result`. Since the description covers its details.

File: integration_tests/src/main/python/fastparquet_utils.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    for for_remove_key in for_removes:
+        cpu_dict.pop(for_remove_key)
+    row = Row(**stripped_dict)
+    # trim the end dot, e.g.:  "l1.l2.l3." => "l1.l2.l3"
+    prefix_key = cpu_prefix[0 : len(cpu_prefix)-1]
+    # set the merge row
+    cpu_dict[prefix_key] = row
+
+
+def _test_convert_fastparquet_result():
```
Comment (firestarman) [inline integration_tests/src/main/python/fastparquet_utils.py:89]: so, this is only a test, we'd better move it into a `__main__` block to avoid running it whenever this module is imported.

File: integration_tests/src/main/python/fastparquet_utils.py (hunk 3 above)
Comment (res-life) [reply integration_tests/src/main/python/fastparquet_utils.py:51]: gpu hierarchy is the baseline in this method, we try to build a hierarchy according to gpu hierarchy, so passed in `gpu_dict`.

Consider this case:
cpu: Row(a.first = 1, a.second = 2)
gpu: Row(a.first = 1, a.second = 2)
If only pass in cpu without gpu, then converted cpu will be Row(a = Row(first=1, second=2)), In this case gpu actually has column name "a.first", "a.second"

File: integration_tests/src/main/python/asserts.py (hunk 1 above)
Comment (res-life) [reply integration_tests/src/main/python/asserts.py:485]: Good catch

File: integration_tests/src/main/python/fastparquet_utils.py (hunk 3 above)
Comment (firestarman) [reply integration_tests/src/main/python/fastparquet_utils.py:51]: NIT: It is possible to flatten the GPU result instead of reconstruct the CPU rows ? I guess flattening may be simpler than reconstruct the nested rows, and we just only need to compare the converted results.

File: integration_tests/src/main/python/fastparquet_utils.py (hunk 7 above)
Comment (firestarman) [reply integration_tests/src/main/python/fastparquet_utils.py:89]: Sorry, pls ignore the above comment, it is only a function. That is ok

**Comment by res-life** [comment]
>This is true for the fastparquet case here, but may be false for some cases. We'd better not enforce the local sort. Callers should add it when they know it is necessary.

The assert will help developer fast find the failure reason.
If developer use @ignore_order(local=False) and passed in the `result convert func`,  the result will be change, the previous sort no longer take effect,  so need a new sort.

File: integration_tests/src/main/python/fastparquet_utils.py (hunk 7 above)
Comment (res-life) [reply integration_tests/src/main/python/fastparquet_utils.py]: This is valuable.
I put the function into __main__, and now I test it more convenient.
thanks.

File: integration_tests/src/main/python/fastparquet_utils.py (hunk 4 above)
Comment (res-life) [reply integration_tests/src/main/python/fastparquet_utils.py:93]: Added a case for testing null.

File: integration_tests/src/main/python/asserts.py (hunk 6 above)
Comment (res-life) [reply integration_tests/src/main/python/asserts.py]: Let's use `result_canonicalize_func_before_compare`, actually developer may not look at the description, and the parameter name has a good meaning to let developer to know the usage.

**Comment by firestarman** [comment]
> > This is true for the fastparquet case here, but may be false for some cases. We'd better not enforce the local sort. Callers should add it when they know it is necessary.
> 
> The assert will help developer fast find the failure reason. If developer use @ignore_order(local=False) and passed in the `result convert func`, the result will be change, the previous sort no longer take effect, so need a new sort.

If users need the non-local sort, then it will conflict with the requirement of local sort. If local sort is always needed when this conversion func is specfied, we may need to always do the local sort instead of an assertion. And I am not sure, will the order of the output always be changed when giving a conversion func ? This is confusing me and i think it depends on the implementation of this func.

File: integration_tests/src/main/python/fastparquet_utils.py (hunk 3 above)
Comment (res-life) [reply integration_tests/src/main/python/fastparquet_utils.py:51]: In this PR we convert the result to follow the result of Spark. The GPU result also follow the result type of Spark.
I think it's better to take the Spark result as the baseline.

`test_fastparquet` you provided should return a `data frame`,  `assert_gpu_and_cpu_are_equal_collect` expects the function retuns a `df` because `assert_gpu_and_cpu_are_equal_collect` will call `test_fastparquet().collect`.

Comment (firestarman) [reply integration_tests/src/main/python/fastparquet_utils.py:51]: you are right.

Comment (winningsix) [reply integration_tests/src/main/python/fastparquet_utils.py:51]: > d cpu will be Row(a = Row(first=1, second=2)), In this case gpu actually has column name "a.first", "a.second"

This is a case where we can't convert a flatten value back. Sounds like we only care about data value other than the "schema". Then, why not just comparing a canonicalized value directly without rebuilding the tree again?

Comment (res-life) [reply integration_tests/src/main/python/fastparquet_utils.py]: Yes, flatten the GPU result is simple and doable.
Since fastparquet already flattened the result and the schema is missed, the GPU result also can be convert to flatted format.

**Comment by res-life** [comment]
New changes:
- Now flatten the GPU result like fastparquet does to compare, in this way it's more simple and reliable.
- Do not force use @ignore_order(local=False), now we sort it if a canonicalize func is specified.

**Comment by res-life** [comment]
> If users need the non-local sort, then it will conflict with the requirement of local sort. If local sort is always needed when this conversion func is specfied, we may need to forcely do the local sort instead of an assertion. And I am not sure, will the order of the output always be changed when giving a conversion func ? This is confusing me and i think it depends on the implementation of this func.

I tested, first sort on Spark, then convert(do not change order), then sort on local, and the order does not change for the result, so I removed the sort requirement. It's depending caller to sort the data or not.

**Review by mythrocks — APPROVED** [review]

**Comment by firestarman** [comment]
Looks good to me

File: integration_tests/src/main/python/fastparquet_utils.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+from pyspark.sql import Row
+
+def get_fastparquet_result_converter():
```
Comment (winningsix) [inline integration_tests/src/main/python/fastparquet_utils.py:19]: The name may be misleading. It's not returning the convertor but actual the converted result.

Not sure how we plan to use this. Probably you was meaning ```get_fastparquet_canonicalized_result```.

**Review by winningsix — COMMENTED** [review]
LGTM except 1 last nit.

**Review by winningsix — APPROVED** [review]

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
Test passed on my local Spark311,320,330,340,350, so I'll merge it.
Thanks all for your review.
